### PR TITLE
Subclass switching

### PR DESCRIPTION
**Pull request by jacobat**
Another stab at making ARS a little more understandable. The gist of it is that we try to isolate the ActiveRecordShards logic in a class instead of monkey patching `ActiveRecord::Base`.

This PR changes a bunch of things around ActiveRecordShards. First up it drops support for Rails < 5. It also removes the migration parts of ARS as those can be better handled separately. It introduces a new `ActiveRecordShards::ShardedModel` class which sharded models must now inherit from instead of `ActiveRecord::Base`. On the flip side the `not_sharded` instruction is no longer necessary or supported. These changes should make ActiveRecord classes not inheriting from `ActiveRecordShards::ShardedModel` behave more like they would when not using ActiveRecordShards.

This also removes support for the `default_shard` idea as it doesn't make a lot of sense to talk about a default shard. If you want a random shard just use the first shard available.

Additionally this introduces the requirement that a shard must always be selected. It's no longer possible to say `on_shard(nil)`. Only numeric shards are accepted.

Future work on this would be to separate sharding from master/slave logic and move the patches to `ActiveRecord::Base` down to the `ActiveRecordShards::ShardedModel` class making `ActiveRecord::Base` totally ignorant of ActiveRecordShards.

With these changes we'll need to release a new major version of ARS to comply with SemVer.

### CODE REVIEW

File: lib/active_record_shards/default_slave_patches.rb [hunk 1]
```diff
@@ -141,8 +141,6 @@ def model.on_slave_by_default?
           left_reflection.klass.on_slave_by_default?
         end
 
-        # also transfer the sharded-ness of the left table to the join model
-        model.not_sharded unless model.left_reflection.klass.is_sharded?
```
Comment (bquorning) [inline lib/active_record_shards/default_slave_patches.rb]: Will this be a problem?

Comment (jacobat) [reply lib/active_record_shards/default_slave_patches.rb]: Not sure :|

**Comment by grosser** [comment]
I'm a little scared that an inheritance change would break all existing usages and also introduce bugs where things like `base_class` no longer return the right classes ... prefer an include or a `sharded` method

File: lib/active_record_shards/connection_switcher.rb [hunk 2]
```diff
@@ -139,23 +138,35 @@ def current_shard_id
...
-          current_shard_selection.on_slave = options[:slave]
-        end
-
         if options.key?(:shard)
+          unless config = configurations[shard_env]
+            raise "Did not find #{shard_env} in configurations, did you forget to add it to your database config ? (configurations: #{configurations.inspect})"
+          end
+          unless config['shard_names'].include?(options[:shard])
+            raise "Did not find shard #{options[:shard]} in configurations"
+          end
```
Comment (grosser) [inline lib/active_record_shards/connection_switcher.rb]: these validations could go out as their own PR so we can test them quickly and make this PR smaller ?

File: lib/active_record_shards/default_slave_patches.rb (hunk 1 above)
Comment (grosser) [reply lib/active_record_shards/default_slave_patches.rb]: seems super sketchy that we call not_sharded ... maybe this is a 1-off join model ... which means that it should now be `model.sharded if foo` instead ?

**Comment by pschambacher** [comment]
That's is one big and scary PR. Should we test it on one application before merging?

**Comment by jacobat** [comment]
> I'm a little scared that an inheritance change would break all existing usages and also introduce bugs where things like base_class no longer return the right classes ... prefer an include or a sharded method

I considered making it a module one would extend as well. I can give it a shot. `base_class` should still return the right thing though as `ShardedModel` is an abstract class in AR terms.

This will be a new major version and will require changes to existing code so I'm not too worried about breaking existing usage.

**Comment by jacobat** [comment]
> That's is one big and scary PR. Should we test it on one application before merging?

I imagine releasing it as ARS 4.0 beta and test it with the K8s HC branch.

File: lib/active_record_shards/connection_switcher.rb (hunk 2 above)
Comment (jacobat) [reply lib/active_record_shards/connection_switcher.rb]: Sure we can do that.

Comment (bquorning) [reply lib/active_record_shards/connection_switcher.rb]: Except that `options[:shard]` can still be `nil` which cannot be found in config['shard_names']. So the validations need to be tweaked a bit.

File: lib/active_record_shards/sharded_model.rb [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
+      def is_sharded?
+        true
+      end
+    end
+  end
+
   class ShardedModel < ActiveRecord::Base
     self.abstract_class = true
+
+    extend ActiveRecordShards::Ext::ShardedModel
```
Comment (bquorning) [inline lib/active_record_shards/sharded_model.rb]: Why is this better than `def self.is_sharded?` ?

Comment (jacobat) [reply lib/active_record_shards/sharded_model.rb]: You mean why am I extending a module over just defining a method?

Comment (bquorning) [reply lib/active_record_shards/sharded_model.rb]: Yeah. that’s what I mean.

Comment (jacobat) [reply lib/active_record_shards/sharded_model.rb]: I need the module to extend into various ActiveRecord classes where I can't easily change the superclass (models from gems). As I need a module I might as well reuse the module here to keep it slightly more dry.

File: lib/active_record_shards/default_slave_patches.rb (hunk 1 above)
Comment (bquorning) [reply lib/active_record_shards/default_slave_patches.rb]: Can we add a test for this case?

Comment (jacobat) [reply lib/active_record_shards/default_slave_patches.rb]: I've tried adding a test - it's not exactly obvious where to put it though

File: lib/active_record_shards/connection_switcher.rb (hunk 2 above)
Comment (bquorning) [reply lib/active_record_shards/connection_switcher.rb]: Extracted to #143, backport to master in #144.

File: lib/active_record_shards/connection_switcher.rb [hunk 4]
```diff
@@ -144,23 +139,35 @@ def current_shard_id
...
           current_shard_selection.shard = options[:shard]
         end
 
+        if options.key?(:slave)
+          current_shard_selection.on_slave = options[:slave]
```
Comment (bquorning) [inline lib/active_record_shards/connection_switcher.rb]: Should we support master/slave switching without switching shards?

File: lib/active_record_shards/model.rb [hunk 5]
```diff
@@ -3,24 +3,14 @@
...
-        raise "You should only call not_sharded on direct descendants of ActiveRecord::Base"
-      end
-      self.sharded = false
+      ActiveSupport::Deprecation.warn("Calling not_sharded is deprecated. "\
+                                      "Please ensure to still read from the "\
+                                      "account db slave after removing the "\
+                                      "call.")
     end
 
     def is_sharded? # rubocop:disable Naming/PredicateName
```
Comment (bquorning) [inline lib/active_record_shards/model.rb]: Should we seize the moment and rename `.is_sharded?` to `.sharded?`?

**Comment by grosser** [comment]
prefer changes are nicer ... rename in old branch and bump in classic then
fix everything up

**Comment by jacobat** [comment]
I'm not sure what you're saying?

File: lib/active_record_shards/model.rb (hunk 5 above)
Comment (jacobat) [reply lib/active_record_shards/model.rb]: Nope ;)

File: lib/active_record_shards/connection_switcher.rb (hunk 4 above)
Comment (jacobat) [reply lib/active_record_shards/connection_switcher.rb]: I'm not sure what you're getting at?

**Comment by bquorning** [comment]
I think it’s a threading fail. The `…` reveals my original question

> Should we seize the moment and rename `.is_sharded?` to `.sharded?`?

**Comment by jacobat** [comment]
Ah!

File: lib/active_record_shards/connection_switcher.rb (hunk 4 above)
Comment (bquorning) [reply lib/active_record_shards/connection_switcher.rb]: Me neither. This is fine.

**Comment by jacobat** [comment]
I've merged #146 bumping to beta-4.0.0.beta1 on the v4 branch. Does anyone know how to release this onto rubygems.org?

**Comment by bquorning** [comment]
[v4.0.0.beta1](https://rubygems.org/gems/active_record_shards/versions/4.0.0.beta1) has been released 🎉 ✨
